sql/core/src/test/scala/org/apache/spark/sql/connector/WriteDistributionAndOrderingSuite.scala

```diff
@@ -0,0 +1,618 @@
+    checkOrderedDistributionAndSortWithIncompatibleGlobalSort("overwriteDynamic")
+  }
+
+  private def checkOrderedDistributionAndSortWithIncompatibleGlobalSort(command: String): Unit = {
+    val tableOrdering = Array[SortOrder](
+      sort(FieldReference("data"), SortDirection.ASCENDING, NullOrdering.NULLS_FIRST),
+      sort(FieldReference("id"), SortDirection.ASCENDING, NullOrdering.NULLS_FIRST)
+    )
+    val tableDistribution = Distributions.ordered(tableOrdering)
+
+    val writeOrdering = Seq(
+      catalyst.expressions.SortOrder(
+        attr("data"),
+        catalyst.expressions.Ascending,
+        catalyst.expressions.NullsFirst,
+        Seq.empty
+      ),
+      catalyst.expressions.SortOrder(
+        attr("id"),
+        catalyst.expressions.Ascending,
+        catalyst.expressions.NullsFirst,
+        Seq.empty
+      )
+    )
+    val writePartitioning = RangePartitioning(writeOrdering, conf.numShufflePartitions)
+
+    checkWriteRequirements(
+      tableDistribution,
+      tableOrdering,
+      expectedWritePartitioning = writePartitioning,
+      expectedWriteOrdering = writeOrdering,
+      writeTransform = df => df.orderBy(df("data").desc, df("id").asc),
+      writeCommand = command)
+  }
+
+  test("ordered distribution and sort with manual local sort: append") {
+    checkOrderedDistributionAndSortWithManualLocalSort("append")
+  }
+
+  test("ordered distribution and sort with manual local sort: overwrite") {
+    checkOrderedDistributionAndSortWithManualLocalSort("overwrite")
+  }
+
+  test("ordered distribution and sort with manual local sort: overwriteDynamic") {
+    checkOrderedDistributionAndSortWithManualLocalSort("overwriteDynamic")
+  }
+
+  private def checkOrderedDistributionAndSortWithManualLocalSort(command: String): Unit = {
+    val tableOrdering = Array[SortOrder](
+      sort(FieldReference("data"), SortDirection.ASCENDING, NullOrdering.NULLS_FIRST),
+      sort(FieldReference("id"), SortDirection.ASCENDING, NullOrdering.NULLS_FIRST)
+    )
+    val tableDistribution = Distributions.ordered(tableOrdering)
+
+    val writeOrdering = Seq(
+      catalyst.expressions.SortOrder(
+        attr("data"),
+        catalyst.expressions.Ascending,
+        catalyst.expressions.NullsFirst,
+        Seq.empty
+      ),
+      catalyst.expressions.SortOrder(
+        attr("id"),
+        catalyst.expressions.Ascending,
+        catalyst.expressions.NullsFirst,
+        Seq.empty
+      )
+    )
+    val writePartitioning = RangePartitioning(writeOrdering, conf.numShufflePartitions)
+
+    checkWriteRequirements(
+      tableDistribution,
+      tableOrdering,
+      expectedWritePartitioning = writePartitioning,
+      expectedWriteOrdering = writeOrdering,
+      writeTransform = df => df.sortWithinPartitions("data", "id"),
+      writeCommand = command)
+  }
+
+  ignore("ordered distribution and sort with manual repartition: append") {
```

Comment:
I should have added a TODO with a JIRA or remove these tests completely if we don't want to optimize such cases.

The original discussion is [here](https://github.com/apache/spark/pull/29066/files#r535518111).

I think there are cases when we can remove repartition nodes but not in a general case.
For example, `RepartitionByExpr(c1, c2) -> Projects -> RepartitionByExpr(c1, c2)` can be simplified.

Reply:
Moreover, we cannot even dedup cases like `RepartitionByExpr(c1, c2) -> RepartitionByExpr(c1, c2)` as `CollapseRepartitions` runs in the operator optimization batch.

Reply:
I'd be interested to hear what everybody thinks about this.
I see a couple of solutions:
- Move `CollapseRepartition` out from the operator optimization batch.
- Introduce a new optimizer rule similar to `CollapseRepartition` and `EliminateSorts`.
- Introduce a preparation rule like `RemoveRedundantSorts` that would operate on `SparkPlan` nodes.

I guess we better do this in the optimizer so either option 1 or 2 can work.